# Project 1 Jupyter Notebook

# Step 1: Load the data

This project explores U.S. county-level COVID-19 total_cases data. The data was released by White House COVID-19 Team, Joint Coordination Cell, Data Strategy and Execution Workgroup. (https://healthdata.gov/dataset/COVID-19-Community-Profile-Report-County-Level/di4u-7yu6/about_data)

- [📊 Download Dataset (CSV)](COVID-19_Community_Profile_Report_-_County-Level.csv)

In [1]:
import pandas as pd

# Read the CSV
df = pd.read_csv("COVID-19_Community_Profile_Report_-_County-Level.csv")

# Preview the dataset and head 10
print(df.shape)
df.head(10)


(3294, 39)


fips               county state  fema_region                    date  \
0  1000      Unallocated, AL    AL          4.0  05/10/2023 12:00:00 AM   
1  1001   Autauga County, AL    AL          4.0  05/10/2023 12:00:00 AM   
2  1003   Baldwin County, AL    AL          4.0  05/10/2023 12:00:00 AM   
3  1005   Barbour County, AL    AL          4.0  05/10/2023 12:00:00 AM   
4  1007      Bibb County, AL    AL          4.0  05/10/2023 12:00:00 AM   
5  1009    Blount County, AL    AL          4.0  05/10/2023 12:00:00 AM   
6  1011   Bullock County, AL    AL          4.0  05/10/2023 12:00:00 AM   
7  1013    Butler County, AL    AL          4.0  05/10/2023 12:00:00 AM   
8  1015   Calhoun County, AL    AL          4.0  05/10/2023 12:00:00 AM   
9  1017  Chambers County, AL    AL          4.0  05/10/2023 12:00:00 AM   

   cases_last_7_days  cases_per_100k_last_7_days  total_cases  \
0                0.0                         NaN          0.0   
1                8.0                      14.319      19913.0   
2               45.0                      20.158      70521.0   
3                3.0                      12.153       7582.0   
4                4.0                      17.862       8149.0   
5               16.0                      27.669      18872.0   
6                1.0                       9.900       3057.0   
7                2.0                      10.284       6617.0   
8               19.0                      16.725      41931.0   
9                7.0                      21.050      10935.0   

   cases_pct_change_from_prev_week  deaths_last_7_days  ...  \
0                              NaN                 0.0  ...   
1                           -0.333                 0.0  ...   
2                           -0.196                 0.0  ...   
3                           -0.250                 0.0  ...   
4                           -0.556                 0.0  ...   
5                            1.000                 0.0  ...   
6                           -0.750                 0.0  ...   
7                              NaN                 0.0  ...   
8                            0.727                 0.0  ...   
9                            1.333                 0.0  ...   

   pct_icu_beds_used_avg_last_7_days  \
0                                NaN   
1                              0.717   
2                              0.836   
3                              0.750   
4                              0.850   
5                              0.850   
6                              0.717   
7                              0.500   
8                              0.798   
9                              0.717   

   pct_icu_beds_used_abs_change_from_prev_week  \
0                                          NaN   
1                                       -0.150   
2                                       -0.016   
3                                       -0.057   
4                                        0.010   
5                                        0.010   
6                                       -0.150   
7                                        0.381   
8                                        0.008   
9                                        0.055   

   pct_icu_beds_used_covid_avg_last_7_days  \
0                                      NaN   
1                                    0.013   
2                                    0.007   
3                                    0.026   
4                                    0.009   
5                                    0.009   
6                                    0.013   
7                                    0.000   
8                                    0.021   
9                                    0.007   

   pct_icu_beds_used_covid_abs_change_from_prev_week  \
0                                                NaN   
1                                             -0.011   
2                                             -0.009   
3                                           

# Step 2: Identify numeric columns

In [2]:
# Choose which column is worth analyzing and computing next
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3294 entries, 0 to 3293
Data columns (total 39 columns):
 #   Column                                                       Non-Null Count  Dtype  
---  ------                                                       --------------  -----  
 0   fips                                                         3294 non-null   int64  
 1   county                                                       3294 non-null   object 
 2   state                                                        3294 non-null   object 
 3   fema_region                                                  3291 non-null   float64
 4   date                                                         3294 non-null   object 
 5   cases_last_7_days                                            3289 non-null   float64
 6   cases_per_100k_last_7_days                                   3221 non-null   float64
 7   total_cases                                                  3278 non-null   f

# Step 3: Compute mean, median, mode

In [22]:
mean_pandas = df["total_cases"].mean()
median_pandas = df["total_cases"].median()
mode_pandas = df["total_cases"].mode()[0]

print("Mean:", mean_pandas)
print("Median:", median_pandas)
print("Mode:", mode_pandas)

Mean: 31945.17754728493
Median: 8045.5
Mode: 0.0


# Data Visualization

In [23]:
state_cases = df.groupby("state")["total_cases"].sum().sort_values(ascending=False)

print("COVID-19 Total Cases by State by 2023/5/10")
print("(Each 🦠 represents approximately 1,000 total cases)\n")
for state, value in state_cases.items():
    # Adds a cap (min(value // 1000, 100)) so extreme values (like NYC) don’t print 50,000 stars.
    bar = "🦠" * min(int(value) // 100000, 100)
    # Each county still has a proportional bar.
    print(f"{state:5s}{bar}")
    # Use scrollable elements to see full data visualization.

COVID-19 Total Cases by State by 2023/5/10
(Each 🦠 represents approximately 1,000 total cases)

CA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
TX   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
FL   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
NY   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
IL   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
PA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
NC   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
OH   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
MI   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
GA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
NJ   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
TN   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
AZ   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
VA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
MA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
IN   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
WI   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
WA   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
SC   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
MN   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
MO   🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
CO   🦠🦠🦠🦠

# The hard way

In [11]:
import csv
import os

# Computing mean, mode, median
values = []
# Put CSV in same folder as this script
BASE_DIR = os.getcwd()
CSV_PATH = os.path.join(
    BASE_DIR, "COVID-19_Community_Profile_Report_-_County-Level.csv"
)
with open(CSV_PATH, newline="") as f:
    read_data = csv.DictReader(f)
    for row in read_data:
        try:
            value = float(row["total_cases"])
            values.append(value)
        except:
            pass  # skip missing or bad data

# Mean
mean_manual = sum(values) / len(values)

# Median
values_sorted = sorted(values)
n = len(values)
if n % 2 == 1:
    median_manual = values_sorted[n // 2]
else:
    median_manual = (values_sorted[n // 2 - 1] + values_sorted[n // 2]) / 2

# Mode
mode_manual = max(set(values), key=values.count)

print("Mean of total_cases:", mean_manual)
print("Median of total_cases:", median_manual)
print("Mode of total_cases:", mode_manual)


# Data visualization under county-level （Using standard library version)
# Configuration
MAX_BARS = 20  # number of counties to show
BAR_CHAR = "🦠"  # emoji (can change to "*" or anything else)
BAR_WIDTH = 50  # max bar width in characters

# READ DATA
rows = []
with open(CSV_PATH, newline="") as f:
    for r in csv.DictReader(f):
        try:
            # remove commas and spaces, convert total_cases to a number
            v = float(str(r["total_cases"]).replace(",", "").strip())
            if v > 0:  # skip zeros or missing data
                rows.append((r["county"], v))
        except:
            pass  # skip rows that can’t be converted to numbers

# Keep top 20 and scale
rows.sort(key=lambda x: x[1], reverse=True)  # sort by total_cases, descending
rows = rows[:MAX_BARS]  # keep only top 20
max_val = max(v for _, v in rows)  # largest number of cases
unit = max(int(max_val / BAR_WIDTH), 1)  # each emoji represents about how many cases
scale = BAR_WIDTH / max_val if max_val else 1  # how wide each bar should be

# Draw

# Alignment width based on the longest county name
name_w = max(len(name) for name, _ in rows)

print("\nCOVID-19 Total Cases (Top Counties) by 2023/5/10\n")
print(f"Each {BAR_CHAR} ≈ {unit:,} cases\n")

# Print aligned bars, no vertical bar
for name, val in rows:  # loop through each county and its total case count
    n = max(
        1, int(val * scale)
    )  # decide how long the bar should be; at least 1 symbol long
    bar = BAR_CHAR * min(
        n, BAR_WIDTH
    )  # make a string of that many emojis but cap the max width
    print(
        f"{name:<{name_w}}  {bar}"
    )  # print the county name left-aligned, then its bar

# Full data interpretation and analysis is in project1_irina.ipynb

Mean of total_cases: 31945.17754728493
Median of total_cases: 8045.5
Mode of total_cases: 0.0

COVID-19 Total Cases (Top Counties) by 2023/5/10

Each 🦠 ≈ 74,849 cases

Los Angeles County, CA     🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Cook County, IL            🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Miami-Dade County, FL      🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Maricopa County, AZ        🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Harris County, TX          🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
San Diego County, CA       🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Kings County, NY           🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Queens County, NY          🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Orange County, CA          🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Riverside County, CA       🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Broward County, FL         🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
San Bernardino County, CA  🦠🦠🦠🦠🦠🦠🦠🦠🦠🦠
Bexar County, TX           🦠🦠🦠🦠🦠🦠🦠🦠🦠
Dallas County, TX          🦠🦠🦠🦠🦠🦠🦠🦠🦠
Clark County, NV           🦠🦠🦠🦠🦠🦠🦠🦠🦠
Tarrant County, TX         🦠🦠🦠🦠🦠🦠🦠🦠
New York County, NY        🦠🦠🦠🦠🦠🦠🦠🦠
Suffolk County, NY         🦠🦠🦠🦠🦠🦠🦠
King County, WA            🦠🦠🦠🦠🦠🦠🦠
Bronx County, NY           🦠🦠🦠🦠🦠🦠🦠


# Analysis and Data Intepretation

The dataset, “COVID-19 Community Profile Report – County-Level,” contains cumulative COVID-19 case counts by county across the United States.  
The column 'total_cases' represents the total number of confirmed cases recorded for each county since the start of the pandemic to 2023/5/10. The data was released by White House COVID-19 Team, Joint Coordination Cell, Data Strategy and Execution Workgroup. 

## Statistical Summary
- Mean total cases per county: 31945.17754728493
- Median total cases per county: 8045.5
- Mode total cases per county: 0

## Interpretation
The mean is much higher than the median, which suggests a right-skewed distribution.  
CA and TX have extremely high case counts as states, while rural areas have fewer cases. I conduct county data visualization in the standard library, and two results highly align with each other. 
This implies that the pandemic’s burden was unevenly distributed, heavily concentrated in urban areas. Even though the total cases seem large, mode is 0, indicating that some counties/states didn't gather valid data or covid had never happened in that area (which is not very likely).

## Key Assumptions
- Missing or incomplete records were excluded.  
- “Total cases” are influenced by county population size; a more meaningful comparison might normalize by population.

## Conclusion
The project highlights disparities in county-level case distribution and with the help of data visualization in state-wide and county-wide, the statistics can be understood more evidently.
